### PR TITLE
Fix and upgrade regional sampling

### PR DESCRIPTION
**Pull request by johanmazoyer**
This PR is a fix of the regional sampling. Now can achieve 10-10 contrast with high enough parameters. 
- it corrects an error in the shift that was not uniformly done in all the dimension
- It does it better using the X_offset_output and X_offset_input parameters of the mft which allows shift without doing phase ramps
- It introduces value errors in a number of case and remove the outer_sampling parameter that was confusing
Currently include a lot of comments to document what I did, can be cleaned after
- in Asterix most pupil planes are over-padded but the current regional prop function assumed the pupil was exactly the size of the pupil plane array. Added a real_dim_input function, same properties as the one in mft. 
- updated tests

### CODE REVIEW

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:594]: Something missing here

Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py:594]: I've actually corrected that in the other PR so let's ignore

Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py:594]: corrected

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:604]: Could clarify the language:
"List of the number of resolution elements *across* the total image plane for all propagation layers.
As a general rule, it is probably safest to choose these numbers such that there is not sampling shift right in the middle of the DH. This could help in avoiding to confuse the matrix (which matrix?). So if we correct the DH between two radii IWA and OWA (in lambda/D), nbres should not have any elements between 2*IWA and 2*OWA."

Careful, I also fixed some typos here so that's minimum.

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:621]: need -> needs

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:620]: So are arrays allowed for `nbres` or not? You deleted the array as an option from the docstring, so I am confused.

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:629]: You already assigned `dim_overpad_pupil`, so I would use it here instead of `pup.shape[0]`.

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:632]: If possible, please delete this comment, it is just a less cleared repeat of the documentation for `numpy.diff` and almost confusing here.

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:633]: need -> needs
nbrs -> number

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:637]: hardcoded *to* 2

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:646]: I am not sure I understand this block and comment. Could you help me out?

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:653]: Please delete. Commented print statements are very bad coding. If you need it in the future you can add it temporarily, otherwise it just clutters the code.

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:667]: Could we delete this? @rgalicher and I convinced ourselves that this was fine. I am happy to revisit together, but maybe it's not the time to add this comment as if it was a done deal. You could instead write something like: "Shifting needs to be revisited due to confusion about shift units." or whatever you like.

Or, since you changed the way the shifting is done, just delete all this commented code altogether.

File: Asterix/tests/test_coronagraph.py
Comment (ivalaginja) [inline Asterix/tests/test_coronagraph.py:29]: Same like comment above:
Please delete. Commented print statements are very bad coding. If you need it in the future you can add it temporarily, otherwise it just clutters the code.
Especially since print statements are ignored in test functions.

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:577]: Parameter missing in docstring.

File: Asterix/tests/test_propagations.py
Comment (ivalaginja) [inline Asterix/tests/test_propagations.py:165]: regionnal -> regional

File: Asterix/tests/test_propagations.py
Comment (ivalaginja) [inline Asterix/tests/test_propagations.py:166]: No prints in tests please.

File: Asterix/tests/test_testbed.py
Comment (ivalaginja) [inline Asterix/tests/test_testbed.py:42]: 1e-8 or 5e-10? You use both on the same line here.

**Review by ivalaginja — CHANGES_REQUESTED** [review]
I have a couple of questions I would like to clarify before merging this, but nothing major. Apart from that, a couple of cleanup suggestions.

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [reply Asterix/optics/propagation_functions.py:637]: And maybe add a period at the end of the sentence?

File: Asterix/tests/test_coronagraph.py
Comment (johanmazoyer) [reply Asterix/tests/test_coronagraph.py]: I've uncommented it. This is fairly useful to check all coronagraphs best performance and do not appears anyway in test mode only when we run them in the main

File: Asterix/tests/test_propagations.py
Comment (johanmazoyer) [reply Asterix/tests/test_propagations.py]: same comment. This is very useful to check coronagraphs best performance and do not appears anyway in test mode only when we run them in the main. I've changed the comment to be clearer though

File: Asterix/tests/test_testbed.py
Comment (johanmazoyer) [reply Asterix/tests/test_testbed.py]: corrected

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py:573]: the first one is a lot longer and gives worst results so no. 
Anyway, this parameter should probably not be optional, the best number are heavily dependent on the size of the pupil and fpm in pixel, so we cannot just use a default value and hope for the best. I have modified this in the next PR

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: corrected

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py:577]: done

File: Asterix/tests/test_propagations.py
Comment (johanmazoyer) [reply Asterix/tests/test_propagations.py]: done

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [reply Asterix/optics/propagation_functions.py:573]: Understood!

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [reply Asterix/optics/propagation_functions.py]: I just saw that the markdown made by `*` disappear in `2*IWA` and `2*OWA` xD If you feel like it, you could correct this, but it's low priority and I won't block the PR because of this.

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:615]: I think this is not quite correct, the variable `save_all_planes_to_fits` doesn't show up anywhere in here.

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py]: I'll do in the next PR directly. Because I have move this function to another file in the next PR, I need to do all these modification twice because they are not passed on :-D

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:648]: Oh this makes much more sense now, thanks! Just some small fixes:
- 'nbrs' -> 'nbres' (two times in line 647)
- in the case *where* the maximum element...
- correspond -> corresponds

File: Asterix/optics/propagation_functions.py
Comment (ivalaginja) [inline Asterix/optics/propagation_functions.py:653]: Again `nbrs` instead of the variable we actually use, `nbres`. Be careful with these kind of things, it can be very, very confusing. Could you fix it here too? And maybe do a "search for all" in this file and check it's not happening somewhere else as well.

File: Asterix/tests/test_coronagraph.py
Comment (ivalaginja) [reply Asterix/tests/test_coronagraph.py]: Sounds good! Also, I have found this, which clarifies how pytest handles stoud, it actually will print it if the test fails:
https://stackoverflow.com/a/24617813/10112569

**Review by ivalaginja — COMMENTED** [review]
Only some minor comments left.

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py:615]: Oops I copied and past another docstring that was apparently outdated. Changed both

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py:648]: done

File: Asterix/optics/propagation_functions.py
Comment (johanmazoyer) [reply Asterix/optics/propagation_functions.py:653]: done

**Review by ivalaginja — APPROVED** [review]
Good to go! If I may, here is a reading recommendation if you ever have some time to kill on the train... ^^
https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/

Looking at all the "update", "comment", "format", etc. commit messages in here. 👀

**Pull request merged**
